compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs

```diff
@@ -2,67 +2,110 @@
 -- SPDX-License-Identifier: Apache-2.0
 
 module DA.Daml.LF.Reader
-    ( Manifest(..)
-    , ManifestData(..)
-    , manifestFromDar
-    , multiLineContent
-    , getManifestField
+    ( parseManifestFile
+    , readManifest
+    , manifestPath
+    , DalfManifest(..)
+    , Dalfs(..)
+    , readDalfManifest
+    , readDalfs
     ) where
 
 import "zip-archive" Codec.Archive.Zip
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.UTF8 as BSUTF8
 import qualified Data.ByteString.Lazy as BSL
-import qualified Data.ByteString.Lazy.UTF8 as UTF8
-import qualified Data.ByteString.Char8 as BSC
-import qualified Data.HashMap.Strict as Map
+import Data.Char
 import Data.List.Extra
-import System.FilePath
-import Safe
+import Data.Void
+import Text.Megaparsec
+import Text.Megaparsec.Byte
 
-data Manifest = Manifest
-    { mainDalf :: FilePath
-    , dalfs :: [FilePath]
-    } deriving (Show)
+type Parser = Parsec Void ByteString
 
-data ManifestData = ManifestData
-    { mainDalfContent :: BSL.ByteString
-    , dalfsContent :: [BSL.ByteString]
-    } deriving (Show)
+parseManifestFile :: ByteString -> Either String [(ByteString, ByteString)]
+parseManifestFile bs = case parse manifestParser "MANIFEST.MF" bs of
+    Left errBundle -> Left $ errorBundlePretty errBundle
+    Right r -> Right r
+
+readManifest :: Archive -> Either String [(ByteString, ByteString)]
+readManifest dar = do
+    entry <- getEntry dar manifestPath
+    parseManifestFile $ BSL.toStrict entry
+
+manifestPath :: FilePath
+manifestPath = "META-INF/MANIFEST.MF"
+
+-- | We try to be fairly lenient in our parser, e.g., we do not enforce that
+-- lines abide to the 72 byte limit.
+--
+-- See
+-- https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#JAR_Manifest
+-- for a description of the format.
+manifestParser :: Parser [(ByteString, ByteString)]
+manifestParser = do
+    xs <- many manifestHeaderParser
+    _ <- many eol
+    pure xs
 
-lineToKeyValue :: String -> (String, String)
-lineToKeyValue line = case splitOn ":" line of
-    [l, r] -> (trim l , trim r)
-    _ -> error $ "Expected two fields in line " <> line
+manifestHeaderParser :: Parser (ByteString, ByteString)
+manifestHeaderParser = do
+    name <- nameParser
+    _ <- chunk ": "
+    value <- valueParser
+    pure (name, value)
 
-multiLineContent :: String -> [String]
-multiLineContent = filter (not . null) . lines . replace "\n " ""
+nameParser :: Parser ByteString
+nameParser = do
+    x <- alphaNumChar
+    xs <- takeWhileP Nothing (\x -> let xChr = chr $ fromIntegral x in isAlphaNum xChr || xChr == '-' || xChr == '_')
```

Comment:
Maybe this could be factored out / named as `isHeaderChar` to match the spec.

Also, not sure if relevant, but it looks like `isAlphaNum` will accept more than just the ASCII alphanumerics, where the spec defines `alphanum` as `{A-Z} | {a-z} | {0-9}` -- e.g. this rule will accept `É` in the header name.

Reply:
Good idea, I’ve factored out `isHeaderChar`. `isAlphaNum` is fine here even if we want to be strict (which we don’t need to be imho) since we get the Char by applying `chr` to a `Word8`. I’ve added a comment to that effect.

Reply:
I think the `É` example still stands, since it represents the codepoint `U+00C9`. Agreed about not being strict.

Reply:
Oh good point, thanks for pointing it out.